### PR TITLE
Support GroupsAccumulator accumulator for udaf

### LINKED ISSUE #8793: Support `GroupsAccumulator` accumulator for User Defined Functions
### Is your feature request related to a problem or challenge?

Many of DataFusion's built in accumlators use the [`GroupsAccumulator`](https://docs.rs/datafusion/latest/datafusion/physical_expr/trait.GroupsAccumulator.html) interface for a significant performance improvement

See the blog post XXX for more details

However, it is not currently possible to use [`GroupsAccumulator`] with user defined Aggregate functions , which is less than ideal for 2 reasons:
1. User's can't take advantage of the faster interface
2. We can't port built in Aggregate functions to use the same API https://github.com/apache/arrow-datafusion/issues/8708 

### Describe the solution you'd like

I would like to extend the `AggregateUDFImpl` trait added by @guojidan  in https://github.com/apache/arrow-datafusion/pull/8733 with a new method that returns a GroupsAccumulator

1. Add new methods to `AggregateUDFImpl`
2. Wire in the the call into the impl for `AggregateExpr` https://github.com/apache/arrow-datafusion/blob/dd58c5a7b0069e3bade997abad9ec6bed8c8a1c3/datafusion/physical-plan/src/udaf.rs#L75
3. Update `advanced_udaf.rs` example to create a `GroupsAccumulator`

## Adding new methods to `AggregateUDFImpl`:

Following the model of [`PartitionEvaluator`](https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.PartitionEvaluator.html) and [groups_accumulator_supported](https://docs.rs/datafusion/latest/datafusion/physical_expr/trait.AggregateExpr.html#method.groups_accumulator_supported)

```rust
pub trait AggregateUDFImpl {
...
    /// If the aggregate expression has a specialized
    /// [`GroupsAccumulator`] implementation. If this returns true,
    /// `[Self::create_groups_accumulator`] will be called.
    fn groups_accumulator_supported(&self) -> bool {
        false
    }

   /// Return a specialized [`GroupsAccumulator`] that manages state
    /// for all groups.
    ///
    /// For maximum performance, a [`GroupsAccumulator`] should be
    /// implemented in addition to [`Accumulator`].
    fn create_groups_accumulator(&self) -> Result<Box<dyn GroupsAccumulator>> {
        not_impl_err!("GroupsAccumulator hasn't been implemented for {self:?} yet")
    }

}
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by guojidan**
## Which issue does this PR close?

Closes #8793 .

## What changes are included in this PR?

Support GroupsAccumulator accumulator for udaf

### CODE REVIEW

**Comment by guojidan** [comment]
cc @alamb

File: datafusion-examples/examples/advanced_udaf.rs [hunk 1]
```diff
@@ -194,12 +210,196 @@ fn create_context() -> Result<SessionContext> {
     Ok(ctx)
 }
 
+struct GeometricMeanGroupsAccumulator<F>
```
Comment (alamb) [inline datafusion-examples/examples/advanced_udaf.rs:213]:
```suggestion
/// Define a `GroupsAccumulator` for GeometricMean
/// which handles accumulator state for multiple groups at once.
/// This API is significantly more complicated than `Accumulator`, which manages 
/// the state for a single group, but for queries with a large number of groups
/// can be significantly faster. See the `GroupsAccumulator` documentation for
/// more information.
struct GeometricMeanGroupsAccumulator<F>
```

File: datafusion-examples/examples/advanced_udaf.rs [hunk 2]
```diff
@@ -194,12 +210,196 @@ fn create_context() -> Result<SessionContext> {
...
+    /// The type of the internal sum
+    prod_data_type: DataType,
+
+    /// The type of the returned sum
+    return_data_type: DataType,
+
+    /// Count per group (use u64 to make UInt64Array)
+    counts: Vec<u64>,
+
+    /// product per group, stored as the native type
```
Comment (alamb) [inline datafusion-examples/examples/advanced_udaf.rs:230]:
```suggestion
    /// product per group, stored as the native type (not `ScalarValue`)
```

File: datafusion-examples/examples/advanced_udaf.rs [hunk 3]
```diff
@@ -194,12 +210,196 @@ fn create_context() -> Result<SessionContext> {
...
+
+impl<F> GroupsAccumulator for GeometricMeanGroupsAccumulator<F>
+where
+    F: Fn(
+            <Float64Type as ArrowPrimitiveType>::Native,
+            u64,
+        ) -> Result<<Float64Type as ArrowPrimitiveType>::Native>
+        + Send,
+{
+    fn update_batch(
```
Comment (alamb) [inline datafusion-examples/examples/advanced_udaf.rs:268]:
```suggestion
    /// Updates the accumulator state given input. DataFusion provides `group_indices`, the groups that each
    /// row in `values` belongs to as well as an optional filter of which rows passed. 
    fn update_batch(
```

File: datafusion-examples/examples/advanced_udaf.rs [hunk 4]
```diff
@@ -194,12 +210,196 @@ fn create_context() -> Result<SessionContext> {
...
+        total_num_groups: usize,
+    ) -> Result<()> {
+        assert_eq!(values.len(), 1, "single argument to update_batch");
+        let values = values[0].as_primitive::<Float64Type>();
+
+        // increment counts, update sums
+        self.counts.resize(total_num_groups, 0);
+        self.prods
+            .resize(total_num_groups, Float64Type::default_value());
+        self.null_state.accumulate(
```
Comment (alamb) [inline datafusion-examples/examples/advanced_udaf.rs:282]:
```suggestion
        /// Use the `NullState` structure to generate specialized code for null / non null input elements
        self.null_state.accumulate(
```

File: datafusion-examples/examples/advanced_udaf.rs [hunk 5]
```diff
@@ -194,12 +210,196 @@ fn create_context() -> Result<SessionContext> {
...
+                *prod = prod.mul_wrapping(new_value);
+
+                self.counts[group_index] += 1;
+            },
+        );
+
+        Ok(())
+    }
+
+    fn merge_batch(
```
Comment (alamb) [inline datafusion-examples/examples/advanced_udaf.rs:298]:
```suggestion
    /// Merge the results from previous invocations of `evaluate` into this accumulator's state
    fn merge_batch(
```

File: datafusion-examples/examples/advanced_udaf.rs [hunk 6]
```diff
@@ -194,12 +210,196 @@ fn create_context() -> Result<SessionContext> {
...
+            |group_index, new_value: <Float64Type as ArrowPrimitiveType>::Native| {
+                let prod = &mut self.prods[group_index];
+                *prod = prod.mul_wrapping(new_value);
+            },
+        );
+
+        Ok(())
+    }
+
+    fn evaluate(&mut self, emit_to: datafusion_expr::EmitTo) -> Result<ArrayRef> {
```
Comment (alamb) [inline datafusion-examples/examples/advanced_udaf.rs:338]:
```suggestion
    /// Generate output, as specififed by `emit_to` and update the intermediate state
    fn evaluate(&mut self, emit_to: datafusion_expr::EmitTo) -> Result<ArrayRef> {
```

File: datafusion-examples/examples/advanced_udaf.rs [hunk 7]
```diff
@@ -194,12 +210,196 @@ fn create_context() -> Result<SessionContext> {
...
+    counts: Vec<u64>,
+
+    /// product per group, stored as the native type
+    prods: Vec<f64>,
+
+    /// Track nulls in the input / filters
+    null_state: NullState,
+
+    /// Function that computes the final geometric mean (value / count)
+    geo_mean_fn: F,
```
Comment (alamb) [inline datafusion-examples/examples/advanced_udaf.rs:237]: I think the example would be simpler if you removed the generics and simply inlined the definition of `geo_mean_fn` into the callsite in `evaluate`.  The generics are needed for GroupsAccumulators that are specialized on type (e.g. a special one for Float32, Float64, etc).

**Comment by alamb** [comment]
@guojidan  -- thanks again for your work on this PR. I am quite interested in getting this PR merged, so if you don't think you'll have a chance to work on it this week I would be happy to try and help as well

**Comment by guojidan** [comment]
sorry, I have been very busy the other day, but I am come back 😄

**Comment by alamb** [comment]
> sorry, I have been very busy the other day, but I am come back 😄

No worries -- thank you for working on this

File: datafusion/core/tests/user_defined/user_defined_aggregates.rs [hunk 8]
```diff
@@ -621,3 +642,106 @@ impl Accumulator for FirstSelector {
...
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+        Ok(DataType::UInt64)
+    }
+
+    fn accumulator(&self, _arg: &DataType) -> Result<Box<dyn Accumulator>> {
+        // should use groups accumulator
+        panic!("accumulator shouldn't invoke");
```
Comment (alamb) [inline datafusion/core/tests/user_defined/user_defined_aggregates.rs:671]: 👍

**Review by alamb — APPROVED** [review]
Thank you very much @guojidan  -- this looks great

**Pull request merged**